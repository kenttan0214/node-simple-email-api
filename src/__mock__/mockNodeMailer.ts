export const mockCreateTestAccount = jest.fn(() => ({ user: 'test-account', pass: 'test-password' }));
export const mockSendMail = jest.fn();
export const mockCreateTransport = jest.fn(() => ({ sendMail: mockSendMail }));
export const mockGetTestMessageUrl = jest.fn(() => 'https://unit-test-mock-email-url.example.com');

jest.mock('nodemailer', () => ({
  createTestAccount: mockCreateTestAccount,
  createTransport: mockCreateTransport,
  getTestMessageUrl: mockGetTestMessageUrl,
}));
