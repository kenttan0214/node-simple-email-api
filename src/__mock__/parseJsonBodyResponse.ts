// a simple util function to parse the unit test JSON response into Object
export const parseJsonBody = (response: { body: string }): Record<string, any> => {
  return {
    ...response,
    body: JSON.parse(response.body),
  };
};
