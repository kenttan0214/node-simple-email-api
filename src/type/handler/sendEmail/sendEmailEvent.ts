import { APIGatewayProxyEvent } from 'aws-lambda';

export type EmailBody = {
  subject: string;
  content: string;
  recipients: string[];
};

export type SendEmailEvent = {
  body: EmailBody;
} & APIGatewayProxyEvent;
