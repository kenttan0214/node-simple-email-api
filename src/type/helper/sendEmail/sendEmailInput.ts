export type EmailInput = {
  subject: string;
  content: string;
  recipients: string[];
};
