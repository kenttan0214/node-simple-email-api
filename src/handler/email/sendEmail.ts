import { APIGatewayProxyResult } from 'aws-lambda';
import middleware from '../../middleware';
import { SendEmailEvent } from '../../type/handler/sendEmail/sendEmailEvent';
import { badRequest, ok } from '../../util/apiGatewayResponse';
import sendEmailHelper from '../../helper/email/sendEmail';

export const inputSchema = {
  type: 'object',
  required: ['body'],
  properties: {
    body: {
      type: 'object',
      properties: {
        subject: {
          type: 'string',
        },
        content: {
          type: 'string',
        },
        recipients: {
          type: 'array',
          minItems: 1,
          uniqueItems: true,
          items: {
            type: 'string',
            format: 'email',
          },
        },
      },
      required: ['subject', 'content', 'recipients'],
    },
  },
};

const sendEmail = async (event: SendEmailEvent): Promise<APIGatewayProxyResult> => {
  try {
    const {
      body: { subject, content, recipients },
    } = event;

    const response = await sendEmailHelper({ subject, content, recipients });

    return ok(response);
  } catch (error) {
    return badRequest(error);
  }
};

export const handler = middleware(sendEmail, inputSchema);
