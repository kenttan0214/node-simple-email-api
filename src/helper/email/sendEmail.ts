import { createTestAccount, getTestMessageUrl } from 'nodemailer';
import { EmailInput } from '../../type/helper/sendEmail/sendEmailInput';
import smtpTransporter from '../../util/smtpTransporter';

export default async ({ subject, content, recipients }: EmailInput): Promise<{ preview: string | false }> => {
  // use nodemailer test account, actual use case will create own mail acccount such as Gmail
  const testAccount = await createTestAccount();

  // create reusable transporter object using the default SMTP transport
  const transporter = smtpTransporter(testAccount);

  // send mail with defined transport object
  const info = await transporter.sendMail({
    from: '"Fred Foo" <foo@example.com>', // sender address
    to: recipients.toString(), // list of recipients
    subject, // Subject line
    text: content, // plain text body
  });

  return {
    // Preview only available when sending through an Ethereal account
    preview: getTestMessageUrl(info),
  };
};
