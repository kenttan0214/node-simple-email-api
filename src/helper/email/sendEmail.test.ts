import { mockSendMail } from '../../__mock__/mockNodeMailer';
import sendEmail from './sendEmail';

describe('sendEmail', () => {
  it('should trigger transporter.sendMail function with the provided input', async () => {
    await sendEmail({
      subject: 'test email subject',
      content: 'test content',
      recipients: ['baz@example.com', 'bar@example.com'],
    });

    expect(mockSendMail.mock.calls).toMatchSnapshot();
  });

  it('should return the preview URL after success sent mail', async () => {
    const result = await sendEmail({
      subject: 'test email subject',
      content: 'test content',
      recipients: ['baz@example.com', 'bar@example.com'],
    });

    expect(result).toMatchSnapshot();
  });
});
