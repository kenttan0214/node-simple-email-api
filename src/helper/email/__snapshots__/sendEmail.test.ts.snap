// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendEmail should return the preview URL after success sent mail 1`] = `
Object {
  "preview": "https://unit-test-mock-email-url.example.com",
}
`;

exports[`sendEmail should trigger transporter.sendMail function with the provided input 1`] = `
Array [
  Array [
    Object {
      "from": "\\"Fred Foo\\" <foo@example.com>",
      "subject": "test email subject",
      "text": "test content",
      "to": "baz@example.com,bar@example.com",
    },
  ],
]
`;
